### PR TITLE
Token panel

### PR DESCRIPTION
**Pull request by amacar**
[ch1911]

### CODE REVIEW

**Comment by amacar** [comment]
Just a note that after eth-scan library is fixed we need to remove slicing of assets array (https://github.com/MyCryptoHQ/MyCrypto/pull/2742/files#diff-e2aac68b4a380dd99002b79b5b8b9e7cR72).

File: common/v2/features/Dashboard/components/TokenPanel/TokenPanel.tsx [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+          tokens.push(
+            Object.assign(asset, {
+              rate: getRate(asset.ticker as TTicker) || 0,
+              details: existingTokenDetails || {}
+            })
+          );
+        }
+        return tokens;
+      }, []);
```
Comment (vilsbole) [inline common/v2/features/Dashboard/components/TokenPanel/TokenPanel.tsx]: I think we should retrigger the effect at least when `currrentAccounts` changes.

File: common/v2/features/Dashboard/components/TokenPanel/TokenPanel.tsx [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+      );
+
+      // Add token details and token rate info to all assets that have  a contractAddress
+      const tempTokens = totals(selectedAccounts).reduce((tokens: AssetWithDetails[], asset) => {
+        if (asset.contractAddress) {
+          const existingTokenDetails = tokensDetails.find(
+            details => details.address === asset.contractAddress
+          );
+
+          tokens.push(
```
Comment (vilsbole) [inline common/v2/features/Dashboard/components/TokenPanel/TokenPanel.tsx]: Prefer destructuring:
```
return [ ...tokens, Object.assign(asset, {...etc}]
```

File: common/v2/services/ApiService/TokenInfo/TokenInfo.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
+import { AxiosInstance } from 'axios';
+
+import { default as ApiService } from '../ApiService';
+
+let instantiated: boolean = false;
+
+const apiUrl = 'https://api.mycryptoapi.com/tokens';
```
Comment (vilsbole) [inline common/v2/services/ApiService/TokenInfo/TokenInfo.ts]: Use app wide const for url

File: common/v2/services/ApiService/TokenInfo/TokenInfo.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+  constructor() {
+    if (instantiated) {
+      throw new Error(`TokenInfoService has already been instantiated.`);
+    } else {
+      instantiated = true;
+    }
+  }
+
+  public getTokensInfo = async (contractAddresses: string[]) => {
```
Comment (vilsbole) [inline common/v2/services/ApiService/TokenInfo/TokenInfo.ts]: TODO comment to complete later?

File: common/v2/features/Dashboard/components/TokenPanel/AddToken.tsx [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+      <InputField
+        label={translateRaw('ADDRESS')}
+        placeholder={translateRaw('ADD_TOKEN_ADDRESS_PLACEHOLDER')}
+        onChange={e => setAddress(e.target.value)}
+        value={address}
+        inputError={addressError}
+      />
+      <InputField
+        label={translateRaw('TOKEN_DEC')}
+        placeholder={'4'}
```
Comment (vilsbole) [inline common/v2/features/Dashboard/components/TokenPanel/AddToken.tsx]: Prefer default ERC20 18 decimals

File: common/v2/services/ApiService/TokenInfo/TokenInfo.ts (hunk 4 above)
Comment (amacar) [reply common/v2/services/ApiService/TokenInfo/TokenInfo.ts]: Not sure what you mean here. This service is fetching token info and is complete.

File: common/v2/services/ApiService/TokenInfo/TokenInfo.ts (hunk 3 above)
Comment (amacar) [reply common/v2/services/ApiService/TokenInfo/TokenInfo.ts]: Done

File: common/v2/features/Dashboard/components/TokenPanel/TokenPanel.tsx (hunk 1 above)
Comment (amacar) [reply common/v2/features/Dashboard/components/TokenPanel/TokenPanel.tsx]: It is retriggered if accounts or rates is changed (https://github.com/MyCryptoHQ/MyCrypto/pull/2742/commits/ea3272bb8cbee1e945c8eaad0cdf348d1fb88b95#diff-29388158806ca87065463442d4ee4c56L60), but this empty array here is initial value for reduce.

Comment (vilsbole) [reply common/v2/features/Dashboard/components/TokenPanel/TokenPanel.tsx]: mea culpa!

File: common/v2/utils/convert.tsx [hunk 6]
```diff
@@ -1,14 +1,11 @@
 import { bigNumberify, formatEther, BigNumber } from 'ethers/utils';
-import BN from 'bn.js';
 
-// Rate is received as decimal numbers, sometimes with as many as 4 decimals
-// While the balance is a BigNumber.js
-// To operate the conversion we multiple by 10000 (because we may have 4 decimals) in order to do BigNumber math
-// Then we make sure to divide by 10000 to receive the correct Float.
 export const convertToFiat = (balance: BigNumber, rate: number = 1): number => {
```
[the remark this answers is not in the document]
Comment (amacar) [reply common/v2/utils/convert.tsx]: Added few tests

**Pull request merged**